app/services/ems_dashboard_service.rb

```diff
@@ -33,15 +33,36 @@ def recent_resources(model, title, label)
     }.compact
   end
 
+  # Generates the following SQL:
+  #
+  #   SELECT COUNT(id), to_char(date_trunc('day', created_on), 'YYYY-MM-DD')
+  #   FROM   model
+  #   WHERE  ems_id = ?
+  #     AND (created_on > 30.days.ago)
+  #   GROUP BY to_char(date_trunc('day', created_on), 'YYYY-MM-DD')
+  #
+  # Returns in the format:
+  #
+  #   {
+  #     "2000-01-01" => 123,
+  #     "2000-01-02" => 234,
+  #     "2000-01-03" => 345,
+  #     ...
+  #   }
   def recent_records(model)
-    all_records = Hash.new(0)
-    records = model.where('created_on > ? and ems_id = ?', 30.days.ago.utc, @ems.id)
-    records = records.includes(:resource => [:ext_management_system]) if @ems.blank?
-    records.sort_by(&:created_on).uniq.each do |r|
-      date = r.created_on.strftime("%Y-%m-%d")
-      all_records[date] += model.where('created_on = ?', r.created_on).count
-    end
-    all_records
+    db_table     = model.arel_table
+    trunc_args   = [Arel::Nodes::SqlLiteral.new("'day'"), db_table[:created_on]]
+    date_trunc   = Arel::Nodes::NamedFunction.new("date_trunc", trunc_args)
+    to_char_args = [date_trunc, Arel::Nodes::SqlLiteral.new("'YYYY-MM-DD'")]
+    group_by_sql = Arel::Nodes::NamedFunction.new("to_char", to_char_args)
+
+    sql = model.select(:id)
+               .where(:ems_id => @ems.id)
+               .where(db_table[:created_on].gt(30.days.ago.utc))
+               .group(group_by_sql.to_sql)
+    sql = sql.includes(:resource => [:ext_management_system]) if @ems.blank?
```

Comment:
Still unsure if this is needed or necessary...

`"and ems_id = ?"` would have always pointed back to the main model, and never `:ext_management_system`, and we are/were doing a `.count` anyway, so these records would have never been used.  Also, this line would have never been reached, as `@ems.id` from above would have most likely caused a `NoMethodError`, so I am completely unsure whether it has any value.

I only left it in because it was here prior and I didn't want to break anything, but we could simplify this method a bit by removing it completely.

Reply:
@alexander-demichev can you speak to the above at all and why it is written this way?  Couldn't find anything for a rational in your previous PR.

Reply:
@NickLaMuro Hi, I didn't write code to this method, just moved it here from ems_infra_service, to make this common  for cloud and infra, but my bad that I didn't examine it well. :)

Reply:
@alexander-demichev ah, sorry.  I see that now.  By bad.

So I guess this is to you, @h-kataria , since it seems like this was added here:

https://github.com/ManageIQ/manageiq/commit/2498cdb2cbdbd99d5651c5ad7dfbda7834bec45f#diff-482f2c40348265e920c212dd5ba8e8a4R133

(had to do a bit of `git` investigation to find it...)

At that point, it the `"and ems_id = ?"` portion had not been added, and was added as part of a bug fix later, [here](https://github.com/ManageIQ/manageiq-ui-classic/pull/144).  But, I could not find out why the `unless @ems.present?` line was there in the first place.

Do you remember by chance?

Reply:
@NickLaMuro I don't remember much of the history of that line right now, but when viewing a dashboard for a specific provider `@ems` will be set and you would want to get data for selected provider's resources only. On the other hand if you try to view a Container Dashboard `@ems` will not be present, you will need to get all data for objects that were created within the specified time limit. Hope this make sense.

Reply:
I see this includes and I keep thinking we could use the data in memory
[[example]](https://github.com/ManageIQ/manageiq/blob/master/app/models/metric/targets.rb#L109) -- it made a big difference over there - but this is a distraction.

If the question is related to why join to the ems table, then possibly it is joining to ems to ensure that the relationship has not been nulled / the record is archived - But I thought we mark archived using a different mechanism now.

Reply:
if that `count` is valid, then the `select` and `includes` are bogus.
(I tend not to like having a `select` and `count` together - not sure if it will emit a `select(id)` or if it will miss an index)